### ISSUE TITLE
Key generation for method interceptors is not really unique enough

### ISSUE THREAD

**Issue by dsyer** [spring-projects/spring-retry#49]
Things work out most of the time because the default is to have a different local context cache for each interceptor (each `@Retryable`). But if the context cache is shared, we need to make the cache key more unique (it should include the method signature or label as well as the arguments).

**Comment by garyrussell** [comment]
@dsyer FYI, this change causes an issue in spring-amqp. We have a `MissingMessageIdAdvice` which allows an (intended to be occasional) message with a missing messageId to pass through the stateful interceptor - it generates an ID on the fly; such messages actually can be retried once (and only once, based on the redelivered message property). Without that advice, such messages are rejected out of hand.

Anyway, the advice cleans up the cache by removing the key after the retry advice has been invoked. Previously, the key was just the the keyGenerator result, but now it's an array containing the label (or `method.toGenericString()`) and the generated key. So, there is no longer any way for something external to the retry interceptor to manipulate the cache.

I think I can work around it by moving the missing message id logic to the `StatefulRetryOperationsInterceptorFactoryBean` but I thought I should give you a heads-up in case you think this change could affect others, given that the cache is public API.

**Comment by dsyer** [comment]
See comment on that last commit for an alternative (the AMQP interceptor is probably not needed now).

**Comment by garyrussell** [comment]
@dsyer Unfortunately, this is too far down the stack - the keyGenerator is invoked in the retry interceptor which sets the key to an arraylist with the second element `null`.

PR: https://github.com/spring-projects/spring-retry/pull/54 for your perusal.

**Comment by dsyer** [comment]
PR was merged.
